File: backend/src/modules/storage/object-storage.ts
export type StoredObjectMetadata = {
  contentLength: number;
  contentType: string | null;
  etag: string | null;
};

export interface ObjectStorage {
  createUploadUrl(input: {
    key: string;
    contentType: string;
    expiresInSeconds: number;
  }): Promise<string>;
  createDownloadUrl(input: {
    key: string;
    fileName: string;
    expiresInSeconds: number;
  }): Promise<string>;
  headObject(key: string): Promise<StoredObjectMetadata | null>;
  readObjectPrefix(key: string, byteCount: number): Promise<Uint8Array>;
  deleteObject(key: string): Promise<void>;
}

File: backend/src/main/java/vn/edu/uit/careerhub/storage/ObjectStorage.java
package vn.edu.uit.careerhub.storage;

public interface ObjectStorage {
    record Metadata(long contentLength, String contentType, String etag) {}

    String createUploadUrl(String key, String contentType, long expiresInSeconds);
    String createDownloadUrl(String key, String fileName, long expiresInSeconds);
    Metadata headObject(String key);
    byte[] readObjectPrefix(String key, int byteCount);
    void deleteObject(String key);
}
